Initial Commit: Transport SameSite Attribute support

mrsaldana opened the pull request:
#10789

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -94,6 +98,9 @@ http.remoteip.useRemoteIpInAccessLog.desc=If this property is set to true and th
 http.compression.desc=An element that is configured within the httpEndpoint element so that the associated HTTP channel can consider compressing response bodies based on the client's Accept-Encoding header.
 
+http.cookies=Cookie Options
+http.cookies.desc=An element that is configured within the httpEndpoint element so that the associated HTTP channel can consider cookie specific configurations.
```
change
cookie specific
to
cookie-specific

mrsaldana replied:
Changes added

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -216,4 +223,13 @@ http.compression.types=List of content types for response compression
+http.compression.serverPreferredAlgorithm.desc=The configured compression algorithm is used to compress the body of responses when it is specified with a non-zero quality value in the request's Accept-Encoding header. The valid compression algorithms include: deflate, gzip, x-gzip, zlib, and identity.
+
+http.cookies.samesite.lax=Lax cookies
+http.cookies.samesite.lax.desc=List of cookies which should have the value of Lax set to the SameSite attribute if not already defined.
```
the desc sentence is confusing to me, but this is what I think you mean:
List of cookies that should have the SameSite attribute set to a value of "Lax", if not already defined.

mrsaldana replied:
A bit outdated now, changes reviewed and updated.

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -216,4 +223,13 @@ http.compression.types=List of content types for response compression
+http.cookies.samesite.lax.desc=List of cookies which should have the value of Lax set to the SameSite attribute if not already defined.
+
+http.cookies.samesite.none=None cookies
+http.cookies.samesite.none.desc=List of cookies which should have the value of None set to the SameSite attribute if not already defined.
```
same desc edits as above:
List of cookies that should have the SameSite attribute set to a value of "None", if not already defined.

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -216,4 +223,13 @@ http.compression.types=List of content types for response compression
+http.cookies.samesite.none.desc=List of cookies which should have the value of None set to the SameSite attribute if not already defined.
+
+http.cookies.samesite.strict=Strict cookies
+http.cookies.samesite.strict.desc=List of cookies which should have the value of Strict set to the SameSite attribute if not already defined.
```
same desc edits as above:
List of cookies that should have the SameSite attribute set to a value of "Strict", if not already defined.

mrsaldana replied:
Updated messages.

skasund commented:
L2 message review not required

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -216,4 +223,13 @@ http.compression.types=List of content types for response compression
+http.compression.serverPreferredAlgorithm.desc=The configured compression algorithm is used to compress the body of responses when it is specified with a non-zero quality value in the request's Accept-Encoding header. The valid compression algorithms include: deflate, gzip, x-gzip, zlib, and identity.
+
+http.cookies.samesite.lax=List of samesite=lax cookies
+http.cookies.samesite.lax.desc=List of cookie names or patterns that should have the SameSite attribute set to a value of "Lax", if not already defined. A single wildcard (*) character is supported as a standalone value, or following a cookie name prefix. Any cookie name or pattern defined by this list must be unique and not present in the 'sameSiteNone' nor 'sameSiteStrict' configurations.
```
The use of "single" in the second sentence makes me think people might think they only get one wildcard used with a prefix.  What about this (note that I have left the asterisks out only because this editor thinks I mean that the text between them is supposed to be italicized):

A single wildcard ( ) character is supported as a standalone value, or following cookie name prefixes.
or
A single wildcard ( ) character is supported as a standalone value. Any number of cookie name prefixes can have a wildcard as a final character.

(If you think is a good change, make the same edit for the other two .desc's also)

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -164,4 +164,12 @@ compression.unsupportedAlgorithm.useraction= When defining the value, ensure tha
+cookies.samesite.duplicateName.useraction=When defining the 'sameSiteLax', 'sameSiteNone', and 'sameSiteStrict' lists, ensure cookie names and patterns are unique across all three lists.
+
+cookies.samesite.unsupportedWildcard=CWWKT0036W:An unsupported use of the wildcard character was attempted by the value [{0}]. The SameSite configuration will not be set for this value.
+cookies.samesite.unsupportedWildcard.explanation=The wildcard (*) character was used in an unsupported way. Support consists of a single wildcard (*) character, used as a standalone value or following a cookie name prefix.
```
Change second sentence to:
Support consists of a single wildcard (*) character used as a standalone value, or following any number of cookie name prefixes.

jantley-ibm commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -164,4 +164,12 @@ compression.unsupportedAlgorithm.useraction= When defining the value, ensure tha
+
+cookies.samesite.unsupportedWildcard=CWWKT0036W:An unsupported use of the wildcard character was attempted by the value [{0}]. The SameSite configuration will not be set for this value.
+cookies.samesite.unsupportedWildcard.explanation=The wildcard (*) character was used in an unsupported way. Support consists of a single wildcard (*) character, used as a standalone value or following a cookie name prefix.
+cookies.samesite.unsupportedWildcard.useraction=When defining the 'sameSiteLax', 'sameSiteNone', and 'sameSiteStrict' lists, ensure that only a single wildcard (*) character is used per pattern. Additionally, use of this character is supported only under the following two conditions: as a standalone value or following a cookie name prefix. 
```
Change
following a cookie name prefix.
to
following cookie name prefixes.

mrsaldana commented:
#build

mrsaldana commented:
#build

chirp1 commented:
@jantley-ibm @mrsaldana Hi Jeff, I'm looking at this sentence: "List of cookie names or patterns that should have the SameSite attribute set to a value of "Lax", if not already defined.". We avoid using should. Typically, we say "You must do xxx" or "Do xxx" instead of using "should", but I don't know if those two styles apply to what the developer is intending. I also don't know what "if not already defined" means. Does it mean if names or patterns are not already set to the value of Lax, then set them to the value of Lax? I think the sentence can be reworded. Maybe the developer means something like this wording: "List of cookie names and patterns. Each cookie name and cookie pattern must have the SameSite attribute set to the value of Lax.".  .....Lax doesn't need quotes around it since it is a single term and the developer already indicated that Lax is a value. If the value were made of multiple terms like "Business process", then quotes would be needed.

mrsaldana commented:
#build

mrsaldana commented:
@chirp1 @jantley-ibm The reason why "if not already defined" is due to the fact that when the cookie or Set-Cookie header gets to the transport level, the application might have already set a SameSite value. This configuration is meant to be considered only when the cookie/header arrives without SameSite having already been set. If the word "should" is a concern, I propose something along the lines of: 
"List of cookie names or patterns for which the SameSite attribute will be set to a value of Lax, if not already defined."

 I did go ahead and remove the quotations as suggested. We also changed from a <cookie> element to a <samesite> element, so I adjusted the metatype messages accordingly. I'm running a build with these changes, but please @jantley-ibm  if you can check these new suggested changes and re-submit ID approval, I'd greatly appreciate it.

mrsaldana commented:
#build

mrsaldana commented:
#build

jantley-ibm commented:
Thanks for all your help with this one

mrsaldana commented:
#run-libby-bot
